sseft_shap_new.py

```diff
@@ -0,0 +1,189 @@
+def main():
+    print("in main")
+    parser = options.get_validation_parser()
+    args = options.parse_args_and_arch(parser)
+
+    # only override args that are explicitly given on the command line
+    override_parser = options.get_validation_parser()
+    override_args = options.parse_args_and_arch(override_parser, suppress_defaults=True)
+
+    load_models(args)
+    utils.import_user_module(args)
+    use_fp16 = args.fp16
+    use_cuda = torch.cuda.is_available() and not args.cpu
+    load_models(args)
+
+if __name__ == '__main__':
+    main()
+
+# Load_models() gets the model from the fairseq library and sets the checkpoint, it initialises a data iterator and loops over the samples in the dataset. 
+# For now I need to test the masking functions together with the model, later i will adjust the code so that it can handle data from Sound and Vision
+
+
+
+def load_models(args, override_args=None):
+    if override_args is not None:
+            overrides = vars(override_args)
+            overrides.update(eval(getattr(override_args, 'model_overrides', '{}')))
+        else:
+            overrides = None
+        ## Load ensemble and extract model
+        print('| loading model(s) from {}'.format(args.path))
+        models, model_args, task = checkpoint_utils.load_model_ensemble_and_task(
+            [args.path],
+            arg_overrides=overrides,
+        )
+        model = models[0]
+
+        ## Move models to GPU
+        for model in models:
+            if use_fp16:
+                model.half()
+            if use_cuda:
+                model.cuda()
+    print(model_args)
+
+    ## Build criterion
+    criterion = task.build_criterion(model_args)
+    criterion.eval()
+
+    ## Load valid dataset 
+    for subset in args.valid_subset.split(','):
+        try:
+            task.load_dataset(subset, combine=False, epoch=0)
+            dataset = task.dataset(subset)
+        except KeyError:
+            raise Exception('Cannot find dataset: ' + subset)
+
+        # Initialize data iterator
+        itr = task.get_batch_iterator(
+            dataset=dataset,
+            max_tokens=args.max_tokens,
+            max_sentences=args.max_sentences,
+            max_positions=utils.resolve_max_positions(
+                task.max_positions(),
+                *[m.max_positions() for m in models],
+            ),
+            ignore_invalid_inputs=args.skip_invalid_size_inputs_valid_test,
+            required_batch_size_multiple=args.required_batch_size_multiple,
+            seed=args.seed,
+            num_workers=args.num_workers,
+        ).next_epoch_itr(shuffle=False)
+        progress = progress_bar.build_progress_bar(
+            args, itr,
+            prefix='valid on \'{}\' subset'.format(subset),
+            no_progress_bar='simple'
+        )
+
+
+    results_dict = {} ## This dict will gatter the shap values for each data sample key: sample ID value: shap-values for each combination of tokens
```

Comment:
```suggestion
    results_dict = {} ## This dict will gather the shap values for each data sample ({key: sample ID; value: shap-values} for each combination of tokens)
```
It's a bit confusing what the format is (also, if you have multiple results, shouldn't this be an array of key-value pairs?)

Reply:
If you need to keep somewhat complex data, you might want to consider creating a dataclass (https://docs.python.org/3/library/dataclasses.html) to make very explicit what it entails. 
Also, you can use type hints (https://docs.python.org/3/library/typing.html) to keep track of what variables are. Mypy can then automatically check if your types are correct over the flow of your code.

Reply:
Thanks! I will explain @greenw0lf , the dictionary will look like: dict{
key: sample id, value: dict{key: combination id, value: prediction (label), probabilities (array of probs). 
I will test if this is a good format and otherwise look into your tips @Veldhoen!